### ISSUE TITLE
Integrating with OSS-Fuzz

### ISSUE THREAD

**Issue by Google-Autofuzz** [civetweb/civetweb#893]
Greetings civetweb developers and contributors,

We’re reaching out because your project is an important part of the open source ecosystem, and we’d like to invite you to integrate with our [fuzzing](https://www.owasp.org/index.php/Fuzzing) service, [OSS-Fuzz]( https://opensource.googleblog.com/2016/12/announcing-oss-fuzz-continuous-fuzzing.html ). OSS-Fuzz is a free fuzzing infrastructure you can use to identify security vulnerabilities and stability bugs in your project. OSS-Fuzz will:

- Continuously run at scale all the fuzzers you write.
- Alert you when it finds issues.
- Automatically close issues after they’ve been fixed by a commit.
- Provide fuzzing on pull requests through optional [CI-Fuzz integration](https://google.github.io/oss-fuzz/getting-started/continuous-integration/).

Many widely used [open source projects]( https://github.com/google/oss-fuzz/tree/master/projects ) like OpenSSL, FFmpeg, LibreOffice, and ImageMagick are fuzzing via OSS-Fuzz, which helps them find and remediate [critical issues]( https://bugs.chromium.org/p/oss-fuzz/issues/list?can=1&q=status%3AFixed%2CVerified+Type%3ABug%2CBug-Security+-component%3AInfra+ ). 

Even though typical integrations can be done in < 100 LoC, we have a [reward program]( https://www.google.com/about/appsecurity/patch-rewards/ ) in place which aims to recognize folks who are not just contributing to open source, but are also working hard to make it more secure.

We want to stress that anyone who meets the eligibility criteria and integrates a project with OSS-Fuzz is eligible for a reward.

To help you getting started, we can write a fuzzer for your project that you are welcome to use directly, or to use it as a starting point.

If you're not interested in integrating with OSS-Fuzz, it would be helpful for us to understand why—lack of interest, lack of time, or something else—so we can better support projects like yours in the future.

If we’ve missed your question in our [FAQ]( https://google.github.io/oss-fuzz/faq/ ), feel free to reply or reach out to us at oss-fuzz-outreach@googlegroups.com.

Thanks!

Tommy
OSS-Fuzz Team

**Comment by bel2125** [comment]
Thank you for your invitation, we accept with thanks.
There is a first draft for some fuzz targets in civetweb: https://github.com/civetweb/civetweb/tree/master/fuzztest
Probably some fine tuning is still required, before it can run on a hosted infrastructure.
According to some FAQ, a pull request with a civetweb folder for https://github.com/google/oss-fuzz/tree/master/projects should be submitted - probably I will need some help there.

**Comment by bel2125** [comment]
@xtne6f : Do you want your email address (you used in your git commits) to be added to the OSS-Fuzz project definition file, so you might get an email if the test finds some issues?

```
homepage: "https://civetweb.github.io/civetweb/"
language: c
primary_contact: "bel2125@gmail.com"
sanitizers:
  - address  (???)
  - undefined  (???)
auto_ccs:
  - (add email ???)
```

**Comment by xtne6f** [comment]
Thank you. Please add my address.

**Comment by DavidKorczynski** [comment]
Hi all!

I did the initial integration on the OSS-Fuzz side of things, so we could get this completed

The PR is here: https://github.com/google/oss-fuzz/pull/4983

@bel2125 @xtne6f are you still interested in this? In that case can I put your emails on the `project.yaml` on the OSS-Fuzz repository?

**Comment by bel2125** [comment]
Yes, definitely still interested.
I already tried to prepare some pull request for google/oss-fuzz, but did not complete it myself, because this infrastructure needs some time to understand, so I postponed it. I did not expect that you (someone else) would be willing to create a first pull request - thank you.
This is what I thought that would be required in the projects/civetweb directory, including a project.yaml with emails.
[civetweb.zip](https://github.com/civetweb/civetweb/files/5882978/civetweb.zip)
I tried to create some tests in the civetweb repository and some build script - they are here in the civetweb repository, together with a README.md: https://github.com/civetweb/civetweb/tree/master/fuzztest
I just had no idea how to test this (other than locally on a Linux system).

**Comment by DavidKorczynski** [comment]
Awesome! I have updated the repository over here with your emails https://github.com/google/oss-fuzz/pull/4983

The CI will complain because of some issues with closed ports, but I discussed this issue with the maintainers and I think they are happy to integrate it in. Once that PR is merged you should start receiving some results in your email inbox when bugs happen.

And yup, I saw the fuzzing set up you had created and the integration with OSS-Fuzz I did uses your fuzzers!

**Comment by bel2125** [comment]
> Awesome! I have updated the repository over here with your emails google/oss-fuzz#4983

Thank you. Is the remaining "@adalogics.com" email in auto_css intentional or a leftover from the template?

Is there anything else I need to do?

**Comment by DavidKorczynski** [comment]
Ah sorry - that is my email. I added this email because if errors happen in the beginning and OSS-Fuzz complains, then I will get the notifications and fix them so OSS-Fuzz runs smooth. I am happy to remove my email, but I would recommend leaving it a bit and then removing it later.

**Comment by bel2125** [comment]
Feel free to leave your email there.
I just wanted to make sure nobody unrelated to CivetWeb complains about "spam emails" from the fuzz test environment. 

The options file in my zip and the additional libraries (build-essential cmake libssl-dev - in addition to zlib1g-dev) are not required? I think I read this in some documentation or sample in summer ... but, as already said, I did not have the time to deal with all the details of this test infrastructure. So you probably know better what settings to use.

> Nothing more needed from you at the moment.

Will there be any email once the setup is complete? Then we can use this as "everything is set up" signal and close the issue.

Thanks again.

**Comment by DavidKorczynski** [comment]
Awesome.

The only package I had to install was zlib1g-dev, which is taken care of in the Dockerfile. The fuzzers run smooth! 

There won't be any email once the it's complete, but once it is complete you will be able to login by way of https://oss-fuzz.com and watch statistics, so when that happens we could close this issue?

**Comment by DavidKorczynski** [comment]
@bel2125 the fuzzers have been merged in now! There seems to be some minor issues with not being able to bind the necessary ports on the servers, however. I will take a look at this a little bit later, but you can leave that with me! As such, it seems this is issue is somewhat good to be closed?

**Comment by bel2125** [comment]
>  I will take a look at this a little bit later, but you can leave that with me!

Thank you.
I already had a brief look and there are some things I don't understand:
Why is the fuzzer using "civetweb_fuzz3" instead of fuzz1 or fuzz2? The first two are tests for the web server, the third is a test for civetweb as a client. While civetweb has strong features as a web server, the client interface is kept minimal. The server would be the main use case, the client is just an addon e.g. when the server is used as a proxy.

https://github.com/civetweb/civetweb/blob/d2d2ea4bd4399c8105c0fe5106eb47b2f748a938/fuzztest/README.txt#L6-L19

The problem in the client test (fuzz3) seems to be it is not possible to bind to port 8080 inside the fuzzing environment. This port is used by the "mock server" that will send the fuzz data to the civetweb client.
However, not being able to bind to port 8080 is a problem also for fuzz1 and fuzz2.

Sooner or later, there will be fuzz4 for the HTTP2 server.
Is there an option to run all 4 tests?

**Comment by bel2125** [comment]
After changing the port to PORT_NUM_HTTP=8081 and adding some diagnostic traces we still have the same problem:

https://github.com/civetweb/civetweb/blob/cb38d861f8032fb194f3bca11d6392354ffb88f8/fuzztest/fuzzmain.c#L149-L157

The log says, we cannot bind to this port either:
```
Step #5: Error: Cannot bind [Address already in use]

Step #5: !!! aborting fuzz test in line 157 !!!AddressSanitizer:DEADLYSIGNAL
```
I addad a loop to try all ports from 1024 to 65535 for `civetweb_fuzz3`- this seems to work ... at least some ports are free to use.
With this we have our first successful oss-fuzz run for `civetweb_fuzz3`.

@DavidKorczynski : The definition to use `civetweb_fuzz3` seems to be here:
https://github.com/google/oss-fuzz/blob/b68e5c0f9a6dc2844a8a92b9fba09948d06b7462/projects/civetweb/build.sh#L21-L22
Is it possible/feasible to run multiple fuzz targets here? E.g., to use "build_and_run.sh" instead of "build.sh" from https://github.com/civetweb/civetweb/tree/master/fuzztest ? From my first impression, this is not intended in the oss-fuzz environment.
Then we should probably swap the oss-fuzz target.

**Comment by DavidKorczynski** [comment]
@bel2125 apologies for the late reply. I worked on this to try and get some fixes up but there were issues with respect to `civetweb_fuzz1` and `civetweb_fuzz2`. 

The way it works is that in the `build.sh` script (the one you link to) we need to move all statically-linked fuzz targets to the `$OUT` folder in the docker container. Then, Google Clusterfuzz will grab these and run them. As such, we shouldn't use `build_and_run.sh` but rather copy all of the `civetweb_fuzzX` files out to the `$OUT` folder. 

However, I am having some troubles with `civetweb_fuzz1` and `civetweb_fuzz2` at the moment. Let me just do a PR on OSS-Fuzz where we enable these fuzzers and then we can see the behavior on the CI machines.

**Comment by DavidKorczynski** [comment]
When running these through OSS-Fuzz, I get:
For `civetweb_fuzz1`:
```
CivetWeb test server failed to start
!!! aborting fuzz test in line 56 !!!sh: 1: netstat: not found
AddressSanitizer:DEADLYSIGNAL
=================================================================
==12==ERROR: AddressSanitizer: ABRT on unknown address 0x00000000000c (pc 0x7f0c57e2e438 bp 0x7fff5a00f410 sp 0x7fff5a00f268 T0)
SCARINESS: 10 (signal)
    #0 0x7f0c57e2e438 in raise (/lib/x86_64-linux-gnu/libc.so.6+0x35438)
    #1 0x7f0c57e30039 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x37039)
    #2 0x5bada4 in init_civetweb /src/civetweb/fuzztest/fuzzmain.c:56:3
    #3 0x5bab03 in LLVMFuzzerTestOneInput_URI /src/civetweb/fuzztest/fuzzmain.c:304:3
```

For `civetweb_fuzz2`
```
CivetWeb test server failed to start
!!! aborting fuzz test in line 56 !!!sh: 1: netstat: not found
AddressSanitizer:DEADLYSIGNAL
=================================================================
==13==ERROR: AddressSanitizer: ABRT on unknown address 0x00000000000d (pc 0x7fd6ced73438 bp 0x7ffe8b9d8630 sp 0x7ffe8b9d8488 T0)
SCARINESS: 10 (signal)
    #0 0x7fd6ced73438 in raise (/lib/x86_64-linux-gnu/libc.so.6+0x35438)
    #1 0x7fd6ced75039 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x37039)
    #2 0x5bb124 in init_civetweb /src/civetweb/fuzztest/fuzzmain.c:56:3
    #3 0x5baba0 in LLVMFuzzerTestOneInput_REQUEST /src/civetweb/fuzztest/fuzzmain.c:323:3

```

**Comment by bel2125** [comment]
Could we just do `mv civetweb_fuzz* $OUT/` ... once HTTP/2 has it's own fuzz target, there will be civetweb_fuzz4 ... and maybe others will follow.


---


' !!! aborting fuzz test in line 56 !!'

https://github.com/civetweb/civetweb/blob/23c9533865dd4a4c0b41a27aa107ef57b5181099/fuzztest/fuzzmain.c#L52-L56

I think this is the same as with fuzz3: Port 8081 is not available, so we cannot start the server. 
For fuzz3 I'm calling `bind` in a loop until I find a suitable port (or all >60000 bind attempts failed).
We "could" try this loop with mg_start as well, but we "should" choose port 0, let the system find a port and pass this port number to the client. That's a little bit more to implement in fuzzmain.c - I can do this later today.

**Comment by DavidKorczynski** [comment]
Okay, let me know once you have this up, then I will put in the `mv civetweb_fuzz* $OUT/` change so we also trigger the CI to run.

**Comment by bel2125** [comment]
Commit "fuzztest: let the fuzz container choose the server port" (https://github.com/civetweb/civetweb/commit/92c7c4e9434f84e0e3b0cec5d879822d619ae85b) should do this for fuzz1 and fuzz2.
The test is running and will take about an hour to complete.

**Comment by bel2125** [comment]
Since OSS-fuzz is running since quite some time for civetweb, we can close this issue here.

see https://oss-fuzz-build-logs.storage.googleapis.com/index.html#civetweb
